@types/cucumber

Authors: @abraaoalves, @jan-molak, @isoung, @BendingBender

---

`StepDefinitionParam` is [hard-coded to a small set of possible types](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/e4ce585bf2771f3913a1f75d5c50982658938748/types/cucumber/index.d.ts#L22), including `string` and `number`. However, it is possible for a custom parameter type to have been registered with `defineParameterType`, which may return a different type to pass into the `StepDefinitionCode`. Right now extra casts are needed to make this work:

```ts
// See also DefinitelyTyped/DefinitelyTyped#19548
defineParameterType({typeName: "list", regexp: /".*?"(?:,\s*".*?")+/, transformer: (rawList: string) => {
  return rawList.split(/,\s*/);
}

// The inputs parameter is actually a string[] but casts are needed
When("my inputs are set to {list}", (inputsParam: any) => {
  const inputs: string[] = inputsParam;
  // set the inputs
});
```

I'd love it if it were somehow possible to get the correct type for the parameter without casts. I've tried declaring the additional types myself, for example like this:

```ts
declare module "cucumber" {
	export type StepDefinitionParam = string | string[] | number | CallbackStepDefinition | TableDefinition;
}
```

However this doesn't work as TypeScript doesn't allow me to override the type alias in this way. I don't know enough about type declaration to suggest a better course of action. Widening `StepDefinitionParam` doesn't seem ideal either, as you lose autocomplete and checking for the standard parameter types.

Is there any better way to get custom registered parameter types into `StepDefinitionParam`?

With the release of TypeScript 2.6 this has gotten even more complicated: when transpiled in with option `--strictFunctionTypes`, even step definition code with e.g. string params throws errors like these:

```
my.step.ts(15,48): error TS2345: Argument of type '(this: World, param: string) => void' is not assignable to parameter of type 'StepDefinitionCode'.
  Types of parameters 'param' and 'stepArgs' are incompatible.
    Type 'StepDefinitionParam' is not assignable to type 'string'.
      Type 'number' is not assignable to type 'string'.
```

--- a/types/cucumber/index.d.ts
+++ b/types/cucumber/index.d.ts
@@ -23,9 +23,7 @@ export interface TableDefinition {
     hashes(): Array<{ [colName: string]: string }>;
 }
 
-export type StepDefinitionParam = string | number | CallbackStepDefinition | TableDefinition;
-
-export type StepDefinitionCode = (this: World, ...stepArgs: StepDefinitionParam[]) => PromiseLike<any> | any | void;
+export type StepDefinitionCode = (this: World, ...stepArgs: any[]) => PromiseLike<any> | any | void;
 
 export interface StepDefinitionOptions {
     timeout?: number;

--- a/types/cucumber/index.d.ts
+++ b/types/cucumber/index.d.ts
@@ -1,4 +1,4 @@
-// Type definitions for cucumber-js 2.0
+// Type definitions for cucumber-js 2.1
 // Project: https://github.com/cucumber/cucumber-js
 // Definitions by: Abraão Alves <https://github.com/abraaoalves>
 //                 Jan Molak <https://github.com/jan-molak>

--- a/types/cucumber/index.d.ts
+++ b/types/cucumber/index.d.ts
@@ -23,7 +23,7 @@ export interface TableDefinition {
     hashes(): Array<{ [colName: string]: string }>;
 }
 
-export type StepDefinitionCode = (this: World, ...stepArgs: any[]) => PromiseLike<any> | any | void;
+export type StepDefinitionCode = (this: World, ...stepArgs: any[]) => any;
 
 export interface StepDefinitionOptions {
     timeout?: number;

